File: go/src/leetcode/offerII/Offer42.go
package offerII

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/16 14:06
 */

type RecentCounter struct {
	nums []int
}

func ConstructorRecentCounter() RecentCounter {
	return RecentCounter{
		nums: make([]int, 0),
	}
}

func (this *RecentCounter) Ping(t int) int {
	this.nums = append(this.nums, t)
	for (this.nums)[0] < t-3000 {
		this.nums = this.nums[1:]
	}
	return len(this.nums)
}

File: java/src/leetcode/offer/Offer42.java
package leetcode.offer;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/9/13 00:53
 */
public class Offer42 {

    public int maxSubArray(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        int[] dp = new int[nums.length];
        dp[0] = nums[0];
        int max = nums[0];
        for (int i = 1; i < nums.length; i++) {
            dp[i] = Math.max(0, nums[i]) + dp[i - 1];
            max = Math.max(max, dp[i]);
        }
        return max;
    }
}
